src/target/riscv/riscv.c

```diff
@@ -480,10 +480,157 @@ static int riscv_create_target(struct target *target, Jim_Interp *interp)
 	return ERROR_OK;
 }
 
+static struct jim_nvp nvp_ebreak_config_opts[] = {
+	{ .name = "m", .value = RISCV_DCSR_EBREAK_M_CTL },
+	{ .name = "s", .value = RISCV_DCSR_EBREAK_S_CTL },
+	{ .name = "u", .value = RISCV_DCSR_EBREAK_U_CTL },
+	{ .name = "vs", .value = RISCV_DCSR_EBREAK_VS_CTL },
+	{ .name = "vu", .value = RISCV_DCSR_EBREAK_VU_CTL },
+	{ .name = NULL, .value = N_RISCV_DCSR_EBREAK_CTL }
+};
+
+#define RISCV_EBREAK_MODE_INVALID -1
+
+static struct jim_nvp nvp_ebreak_mode_opts[] = {
+	{ .name = "exception", .value = false },
+	{ .name = "halt", .value = true },
+	{ .name = NULL, .value = RISCV_EBREAK_MODE_INVALID }
+};
+
+static int jim_configure_ebreak(struct target *target, struct jim_getopt_info *goi)
+{
+	struct riscv_private_config * const pc = target->private_config;
+	struct jim_nvp *common_mode_nvp;
+	if (jim_nvp_name2value_obj(goi->interp, nvp_ebreak_mode_opts, goi->argv[0],
+				&common_mode_nvp) == JIM_OK) {
+		/* Here a common "ebreak" action is processed, e.g:
+		 * "riscv.cpu configure -ebreak halt"
+		 */
+		for (int ebreak_ctl_i = 0; ebreak_ctl_i < N_RISCV_DCSR_EBREAK_CTL; ++ebreak_ctl_i)
+			pc->dcsr_ebreak_ctls[ebreak_ctl_i] = common_mode_nvp->value;
+		return jim_getopt_obj(goi, NULL);
+	}
+
+	/* Here a "ebreak" action for a specific execution mode is processed, e.g:
+	 * "riscv.cpu configure -ebreak m halt"
+	 */
+	struct jim_nvp *ctrl_nvp;
+	if (jim_getopt_nvp(goi, nvp_ebreak_config_opts, &ctrl_nvp) != JIM_OK) {
+		jim_getopt_nvp_unknown(goi, nvp_ebreak_config_opts, /*hadprefix*/ true);
+		return JIM_ERR;
+	}
+	struct jim_nvp *mode_nvp;
+	if (jim_getopt_nvp(goi, nvp_ebreak_mode_opts, &mode_nvp) != JIM_OK) {
+		jim_getopt_nvp_unknown(goi, nvp_ebreak_mode_opts, /*hadprefix*/ true);
+		return JIM_ERR;
+	}
+	pc->dcsr_ebreak_ctls[ctrl_nvp->value] = mode_nvp->value;
+	return JIM_OK;
+}
+
+static int str_ebreak_config(struct target *target, char *buffer)
+{
+	int len = 0;
+	const char *separator = "";
+	for (int ebreak_ctl_i = 0; ebreak_ctl_i < N_RISCV_DCSR_EBREAK_CTL;
+			++ebreak_ctl_i) {
+		const char * const format = "%s%s %s";
+		const char * const ctl = jim_nvp_value2name_simple(nvp_ebreak_config_opts, ebreak_ctl_i)->name;
+		const struct riscv_private_config * const pc = riscv_private_config(target);
+		const char * const mode = jim_nvp_value2name_simple(nvp_ebreak_mode_opts,
+				pc->dcsr_ebreak_ctls[ebreak_ctl_i])->name;
+		if (!buffer)
+			len += snprintf(NULL, 0, format, separator, ctl, mode);
+		else
+			len += sprintf(buffer + len, format, separator, ctl, mode);
+
+		separator = " ";
+	}
+	return len;
+}
+
+static int jim_report_ebreak_config(struct target *target, Jim_Interp *interp)
+{
+	const int len = str_ebreak_config(target, NULL);
+	char *str = malloc(len + 1);
+	if (!str) {
+		LOG_ERROR("Unable to allocate a string of %d bytes.", len + 1);
+		return JIM_ERR;
+	}
+	str_ebreak_config(target, str);
+	Jim_SetResultString(interp, str, len);
+	free(str);
+	return JIM_OK;
+}
```

Comment:
The number of arguments (`goi->argc`) could be checked in this function, too.

Reply:
Is this necessary?
This handler does not require arguments (`goi` is not passed to this function).

Reply:
I take my comment back. By trying it out, I have just found out that OpenOCD already provides proper validation for the arguments of `<target_name> cget`:

```
> my_target cget -nonsense     
Unknown param: -nonsense, try one of: -type, -event, -work-area-virt, -work-area-phys, -work-area-size, -work-area-backup, -endian, -coreid, -chain-position, -dbgbase, -rtos, -defer-examine, -gdb-port, or -gdb-max-connections
> my_target cget -ebreak -nonsense
Unknown param: -nonsense, try one of: -type, -event, -work-area-virt, -work-area-phys, -work-area-size, -work-area-backup, -endian, -coreid, -chain-position, -dbgbase, -rtos, -defer-examine, -gdb-port, or -gdb-max-connections
> my_target cget -ebreak -event   
wrong # args: should be "-event ?event-name? ..."

```

As a general rule, my preference would be to perform strict validation of any user's input. Of course only in case the generic OpenOCD code does not perform the validation for us already.

Reply:
Addressed.